polkadot/xcm/xcm-executor/src/lib.rs

```diff
@@ -649,25 +704,32 @@ impl<Config: config::Config> XcmExecutor<Config> {
 				Ok(())
 			},
 			InitiateTeleport { assets, dest, xcm } => {
-				// We must do this first in order to resolve wildcards.
-				let assets = self.holding.saturating_take(assets);
-				for asset in assets.assets_iter() {
-					// We should check that the asset can actually be teleported out (for this to
-					// be in error, there would need to be an accounting violation by ourselves,
-					// so it's unlikely, but we don't want to allow that kind of bug to leak into
-					// a trusted chain.
-					Config::AssetTransactor::can_check_out(&dest, &asset, &self.context)?;
-				}
-				for asset in assets.assets_iter() {
-					Config::AssetTransactor::check_out(&dest, &asset, &self.context);
+				let old_holding = self.holding.clone();
+				let result = Config::TransactionalProcessor::process(|| -> Result<(), XcmError> {
+					// We must do this first in order to resolve wildcards.
+					let assets = self.holding.saturating_take(assets);
+					for asset in assets.assets_iter() {
+						// We should check that the asset can actually be teleported out (for this
+						// to be in error, there would need to be an accounting violation by
+						// ourselves, so it's unlikely, but we don't want to allow that kind of bug
+						// to leak into a trusted chain.
+						Config::AssetTransactor::can_check_out(&dest, &asset, &self.context)?;
+					}
+					for asset in assets.assets_iter() {
+						Config::AssetTransactor::check_out(&dest, &asset, &self.context);
+					}
+					// Note that we pass `None` as `maybe_failed_bin` and drop any assets which
+					// cannot be reanchored  because we have already checked all assets out.
+					let assets = Self::reanchored(assets, &dest, None);
+					let mut message = vec![ReceiveTeleportedAsset(assets), ClearOrigin];
+					message.extend(xcm.0.into_iter());
+					self.send(dest, Xcm(message), FeeReason::InitiateTeleport)?;
```

Comment:
Once all the `can_check_out` calls pass, then the only thing which can fail is `send`. Therefore if that is done prior to the `check_out` calls, then this can avoid the need for transactionality.

Reply:
We do want to keep the rollback of the holding register, correct?